### PR TITLE
MSLG tool package

### PR DESCRIPTION
**Pull request by feich-ms**
This PR is merging MSLG tool package to master branch.

### CODE REVIEW

File: packages/MSLG/README.MD [hunk 1]
```diff
@@ -0,0 +1,107 @@
+# mslg
```
Comment (vishwacsena) [inline packages/MSLG/README.MD]: To stay consistent with other packages, add this at the top of the readme

> *NOTE: This is a **preview** tool and breaking changes may happen before it is released*

File: packages/MSLG/README.MD [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+## Please notice that this package is still in Preview.
+mslg tool has the ability to parse a .LG file that adheres to the LG file format. CLI tool should report an error on an statically detectable invalid .LG file. MSLG tool should also have the ability to generate a collated .LG file by parsing all provided (or found) .LG files. MSLG tool can also list all possible template expansions for a given template and test data for entity references.
```
Comment (vishwacsena) [inline packages/MSLG/README.MD]: Add link to LG file format -> https://github.com/Microsoft/BotBuilder-Samples/blob/master/experimental/language-generation/lg-file-format.md

Comment (vishwacsena) [inline packages/MSLG/README.MD]: Consider re-writing this to MSLG CLI tool: 
- Reports an error on invalid .lg file. All errors that are statically detectable are reported.
- Can generate a collated .lg file by parsing and analyzing more than one .lg file
- Helps expand any given template (or all templates in a file) based on test values for entity references.

File: packages/MSLG/README.MD [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+- [Node.js](https://nodejs.org/) version 8.5 or higher
+
+## Local Run
+- Install dependency on [botbuilder-lg](https://github.com/Microsoft/botbuilder-js/tree/4.next/libraries/botbuilder-lg/src) package
+  ```
+  1. clone the repo of botbuilder-js
+  2. switch to 4.next branch
+  3. npm install and npm run build
+  ``` 
+  The dependency will be removed once the botbuilder-lg package is published.
```
Comment (vishwacsena) [inline packages/MSLG/README.MD]: Does this comment still apply?

File: packages/MSLG/README.MD [hunk 4]
```diff
@@ -0,0 +1,107 @@
+# mslg
+## Please notice that this package is still in Preview.
+mslg tool has the ability to parse a .LG file that adheres to the LG file format. CLI tool should report an error on an statically detectable invalid .LG file. MSLG tool should also have the ability to generate a collated .LG file by parsing all provided (or found) .LG files. MSLG tool can also list all possible template expansions for a given template and test data for entity references.
+
+## Prerequisite
+
+- [Node.js](https://nodejs.org/) version 8.5 or higher
+
+## Local Run
```
Comment (vishwacsena) [inline packages/MSLG/README.MD]: Consider renaming this to 'Installation'

File: packages/MSLG/src/mslg-parse.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    .option('--stdin', '[Optional] Read .lg file as stream from stdin to validate and collate')
+    .option('--stdout', '[Optional] when set, write out the final file to stdout')
+    .option('--verbose', '[Optional] Flag option used to request verbose output. With this option set, additional useful parse, validate and collate logs are written to stdout')
+    .option('-c, --collate', '[Optional] If not set, same template name across multiple .lg files will throw exceptions.')
+    .parse(process.argv);
+   
+if (process.argv.length < 3) {
+    program.help();
+} else {
+    let parser: any = new Parser();
```
Comment (vishwacsena) [inline packages/MSLG/src/mslg-parse.ts]: consider wrapping in try catch with a good description of the error/ exception

File: packages/MSLG/src/mslg-expand.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    .option('--all', '[Optional] Flag option to request that all templates in the .lg file be expanded.')
+    .option('-i, --interactive', '[Optional] Flag option to request that all missing entity value references be obtained through interactive prompts.')
+    .option('-j, --testInput <testInputJSONFile>', '[Optional] full or relative path to a JSON file containing test input for all variable references.')
+    .option('-o, --out_folder <outputFolder>', '[Optional] Output folder for all files the tool will generate')
+    .parse(process.argv);
+   
+if (process.argv.length < 3) {
+    program.help();
+} else {
+    let expander: any = new Expander();
```
Comment (vishwacsena) [inline packages/MSLG/src/mslg-expand.ts]: Consider wrapping in try.. catch and include a good description of the exception

File: packages/MSLG/src/Parser.ts [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+            let file = filesToParse[0];
+            try {
+                const parseRes = await this.parseFile(file, program.verbose);
+                if (parseRes === undefined) {
+                    return;
+                }
+
+                errors = errors.concat(parseRes);
+            } catch (err) {
+                process.stderr.write(err);
```
Comment (vishwacsena) [inline packages/MSLG/src/Parser.ts]: Its good practice to have this just throw and have the main caller handle it.

File: packages/MSLG/README.MD (hunk 1 above)
Comment (feich-ms) [reply packages/MSLG/README.MD]: fixed

File: packages/MSLG/README.MD (hunk 2 above)
Comment (feich-ms) [reply packages/MSLG/README.MD]: fixed

Comment (feich-ms) [reply packages/MSLG/README.MD]: fixed

File: packages/MSLG/README.MD (hunk 4 above)
Comment (feich-ms) [reply packages/MSLG/README.MD]: fixed

File: packages/MSLG/README.MD (hunk 3 above)
Comment (feich-ms) [reply packages/MSLG/README.MD]: removed

File: packages/MSLG/src/mslg-parse.ts (hunk 5 above)
Comment (feich-ms) [reply packages/MSLG/src/mslg-parse.ts]: fixed and optimized the error handling logic

File: packages/MSLG/src/mslg-expand.ts (hunk 6 above)
Comment (feich-ms) [reply packages/MSLG/src/mslg-expand.ts]: fixed and optimized the error handling logic

File: packages/MSLG/src/Parser.ts (hunk 7 above)
Comment (feich-ms) [reply packages/MSLG/src/Parser.ts]: fixed. Throw all exceptions in parser.ts and expander.ts to called and print them in the same format.

**Comment by feich-ms** [comment]
Hi @gabog why adding blocked label to this PR? Did I missed any background here?

**Pull request merged**